File: ios/SharpReminder/Triggers/RuleSnapshotStore.swift
import Foundation

/// Persistance locale du moteur natif iOS.
///
/// Équivalent de `RuleSnapshotStore.kt`, appuyé sur `UserDefaults` plutôt que
/// sur `SharedPreferences`. Le natif ne lit jamais la base SQLite de
/// l'application : il travaille sur un miroir en lecture seule, poussé par le
/// JavaScript à chaque écriture, et qui doit survivre à la mort du processus.
final class RuleSnapshotStore {

    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    // MARK: - Règles

    func saveRules(_ rawJson: String) {
        defaults.set(rawJson, forKey: Keys.rules)
    }

    func loadRules() -> [RuleSnapshot] {
        guard let raw = defaults.string(forKey: Keys.rules) else { return [] }
        // Un miroir illisible ne doit pas empêcher l'application de démarrer :
        // on repart d'un jeu vide, que la prochaine synchronisation JavaScript
        // reconstruira intégralement.
        return (try? RuleSnapshot.list(fromRawJson: raw)) ?? []
    }

    func removeRule(reminderId: String) {
        guard let raw = defaults.string(forKey: Keys.rules),
              let data = raw.data(using: .utf8),
              let array = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]]
        else { return }

        // Le filtrage opère sur le JSON brut : inutile de reconstruire un
        // format que l'on possède déjà.
        let kept = array.filter { ($0["reminderId"] as? String) != reminderId }
        if let encoded = try? JSONSerialization.data(withJSONObject: kept),
           let string = String(data: encoded, encoding: .utf8) {
            saveRules(string)
        }
    }

    // MARK: - État de satisfaction précédent

    func previousSatisfaction() -> [String: Bool] {
        defaults.dictionary(forKey: Keys.satisfaction) as? [String: Bool] ?? [:]
    }

    func savePreviousSatisfaction(_ states: [String: Bool]) {
        defaults.set(states, forKey: Keys.satisfaction)
    }

    // MARK: - Règles confiées au système

    /// Règles dont la notification a été programmée directement auprès
    /// d'`UNUserNotificationCenter`.
    ///
    /// Sans ce marquage, la prochaine évaluation constaterait la transition
    /// faux → vrai et publierait une **seconde** notification pour un rappel
    /// que le système a déjà délivré, application fermée.
    func osScheduledRuleIds() -> Set<String> {
        Set(defaults.stringArray(forKey: Keys.osScheduled) ?? [])
    }

    func saveOsScheduledRuleIds(_ ids: Set<String>) {
        defaults.set(Array(ids), forKey: Keys.osScheduled)
    }

    // MARK: - File des déclenchements

    func enqueueFiredEvent(_ event: FiredEvent) {
        var queue = defaults.array(forKey: Keys.fired) as? [[String: Any]] ?? []
        queue.append(event.asDictionary)
        defaults.set(queue, forKey: Keys.fired)
    }

    /// Renvoie la file au format JSON et la vide dans la foulée.
    func drainFiredEvents() -> String {
        let queue = defaults.array(forKey: Keys.fired) as? [[String: Any]] ?? []
        defaults.set([], forKey: Keys.fired)

        guard let data = try? JSONSerialization.data(withJSONObject: queue),
              let json = String(data: data, encoding: .utf8)
        else { return "[]" }

        return json
    }

    // MARK: - Diagnostic

    func recordSignal(type: String, at isoInstant: String) {
        var signals = defaults.dictionary(forKey: Keys.lastSignals) as? [String: String] ?? [:]
        signals[type] = isoInstant
        defaults.set(signals, forKey: Keys.lastSignals)
    }

    func lastSignals() -> [String: String] {
        defaults.dictionary(forKey: Keys.lastSignals) as? [String: String] ?? [:]
    }

    private enum Keys {
        static let rules = "sharpReminder.rules"
        static let satisfaction = "sharpReminder.previousSatisfaction"
        static let fired = "sharpReminder.firedEvents"
        static let lastSignals = "sharpReminder.lastSignals"
        static let osScheduled = "sharpReminder.osScheduledRules"
    }
}

File: android/app/src/main/java/com/sharpreminder/triggers/RuleSnapshotStore.kt
package com.sharpreminder.triggers

import android.content.Context
import com.sharpreminder.triggers.model.FiredEvent
import com.sharpreminder.triggers.model.RuleSnapshot
import org.json.JSONArray
import org.json.JSONObject

/**
 * Persistance locale du moteur natif.
 *
 * Le natif ne lit **jamais** la base SQLite de l'application : il travaille sur
 * un miroir en lecture seule, poussé par le JavaScript à chaque écriture. Ce
 * miroir doit survivre à la mort du processus et à un redémarrage du téléphone,
 * d'où `SharedPreferences` — suffisant pour quelques kilo-octets, et disponible
 * sans initialisation depuis un `BroadcastReceiver`.
 *
 * Trois éléments y sont conservés :
 * - les règles actives, telles que reçues ;
 * - l'état de satisfaction précédent de chaque règle, indispensable pour
 *   détecter une transition plutôt qu'un état ;
 * - la file des déclenchements survenus hors ligne, que le JavaScript viendra
 *   vider à son prochain démarrage.
 */
class RuleSnapshotStore(context: Context) {

    private val preferences =
        context.applicationContext.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE)

    // --- Règles -------------------------------------------------------------

    fun saveRules(rawJson: String) {
        preferences.edit().putString(KEY_RULES, rawJson).apply()
    }

    fun loadRules(): List<RuleSnapshot> {
        val raw = preferences.getString(KEY_RULES, null) ?: return emptyList()
        return runCatching { RuleSnapshot.listFromJson(raw) }.getOrElse {
            // Un miroir illisible ne doit pas empêcher l'application de
            // démarrer : on repart d'un jeu vide, que la prochaine
            // synchronisation JavaScript reconstruira intégralement.
            emptyList()
        }
    }

    /**
     * Retire une règle du miroir, après un déclenchement à suppression
     * automatique.
     *
     * Le filtrage opère sur le JSON brut plutôt que sur les objets désérialisés :
     * inutile de reconstruire un format que l'on possède déjà, et cela garantit
     * que ce qui est réécrit reste exactement lisible par `RuleSnapshot.fromJson`.
     */
    fun removeRule(reminderId: String) {
        val raw = preferences.getString(KEY_RULES, "[]") ?: "[]"
        val source = runCatching { JSONArray(raw) }.getOrElse { return }

        val kept = JSONArray()
        for (index in 0 until source.length()) {
            val rule = source.getJSONObject(index)
            if (rule.optString("reminderId") != reminderId) {
                kept.put(rule)
            }
        }

        saveRules(kept.toString())
    }

    // --- État de satisfaction précédent -------------------------------------

    fun previousSatisfaction(): Map<String, Boolean> {
        val raw = preferences.getString(KEY_SATISFACTION, null) ?: return emptyMap()
        return runCatching {
            val json = JSONObject(raw)
            json.keys().asSequence().associateWith { json.getBoolean(it) }
        }.getOrElse { emptyMap() }
    }

    fun savePreviousSatisfaction(states: Map<String, Boolean>) {
        val json = JSONObject()
        states.forEach { (reminderId, satisfied) -> json.put(reminderId, satisfied) }
        preferences.edit().putString(KEY_SATISFACTION, json.toString()).apply()
    }

    // --- File des déclenchements --------------------------------------------

    fun enqueueFiredEvent(event: FiredEvent) {
        val array = JSONArray(preferences.getString(KEY_FIRED, "[]") ?: "[]")
        array.put(event.toJson())
        preferences.edit().putString(KEY_FIRED, array.toString()).apply()
    }

    /** Renvoie la file et la vide dans la foulée. */
    fun drainFiredEvents(): String {
        val raw = preferences.getString(KEY_FIRED, "[]") ?: "[]"
        preferences.edit().putString(KEY_FIRED, "[]").apply()
        return raw
    }

    // --- Diagnostic ---------------------------------------------------------

    fun recordSignal(typeWireName: String, isoInstant: String) {
        val json = JSONObject(preferences.getString(KEY_LAST_SIGNALS, "{}") ?: "{}")
        json.put(typeWireName, isoInstant)
        preferences.edit().putString(KEY_LAST_SIGNALS, json.toString()).apply()
    }

    fun lastSignals(): JSONObject =
        JSONObject(preferences.getString(KEY_LAST_SIGNALS, "{}") ?: "{}")

    private companion object {
        const val PREFERENCES_NAME = "sharp_reminder_triggers"
        const val KEY_RULES = "rules"
        const val KEY_SATISFACTION = "previous_satisfaction"
        const val KEY_FIRED = "fired_events"
        const val KEY_LAST_SIGNALS = "last_signals"
    }
}
